op-challenger: Only respond to claims at levels we disagree with

trianglesphere opened the pull request:
**Description**

This is important to ensure that the challenger would not counter claims that itself made. This also appears to fix the step transaction revert issue.

**Tests**

Unit tests pass. Tested with full game script.



**Metadata**

- Fixes #CLI-4194

ajsutton commented on op-challenger/fault/solver.go:
```diff
@@ -21,14 +21,47 @@ func NewSolver(gameDepth int, traceProvider TraceProvider) *Solver {
+		return nil, err
+	}
+	// Attack the root claim if we do not agree with it
+	// Note: We always do not agree with the level
```
I'm not sure I understand what the Note part of this is trying to say.

trianglesphere replied:
Expanded this comment. It's weird corner case.

ajsutton commented on op-challenger/fault/solver.go:
```diff
@@ -21,14 +21,47 @@ func NewSolver(gameDepth int, traceProvider TraceProvider) *Solver {
+	}
+	// Attack the root claim if we do not agree with it
+	// Note: We always do not agree with the level
+	if !agree {
+		return s.attack(claim)
+	} else {
+		return nil, nil
+	}
```
There doesn't appear to be any tests around this code.

ajsutton commented on op-challenger/fault/solver.go:
```diff
@@ -21,14 +21,47 @@ func NewSolver(gameDepth int, traceProvider TraceProvider) *Solver {
 // NextMove returns the next move to make given the current state of the game.
-func (s *Solver) NextMove(claim Claim) (*Claim, error) {
+func (s *Solver) NextMove(claim Claim, agreeWithClaimLevel bool) (*Claim, error) {
+	if agreeWithClaimLevel {
+		return nil, nil
+	}
```
I can delete this code and no tests fail. :)

trianglesphere replied:
fixed.

ajsutton reviewed (commented):
Probably worth splitting the cannon step fixes out into a separate PR. I'd want someone more familiar with cannon to review them.

trianglesphere commented:
> cannon step fixes out into a separate PR

Sorry, this is a rebase issue. Will re-request when this is a clean diff

trianglesphere committed:
op-challenger: Only respond to claims at levels we disagree with

This is important to ensure that the challenger would not counter claims
that itself made. This also appears to fix the step transaction revert
issue.

refcell reviewed (approved):
LGTM
